File: java/ordenacao/InsertionSort.java
/**
 * Estruturas de Dados e Algoritmos
 * @author Rafael de Arruda Sobral
 */

package ordenacao;

import java.util.Arrays;
import java.util.Scanner;

/**
 * As the insertion sort algorithm iterates over the array, it makes the
 * assumption that the visited positions are already sorted in ascending order,
 * which means it only needs to find the right position for the current element
 * and insert it there.
 */

public class InsertionSort {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String[] input = sc.nextLine().split(" ");
        int[] array = new int[input.length];
        for (int i = 0; i < input.length; i++) {
            array[i] = Integer.parseInt(input[i]);
        }
        insertionSort(array, 0, array.length - 1);
    }

    private static void insertionSort(int[] array, int left, int right) {
        for (int i = left + 1; i <= right; i++) {
            int key = array[i];
            int aux = i - 1;
            while (aux >= left && array[aux] > key) {
                array[aux + 1] = array[aux];
                aux--;
            }
            array[aux + 1] = key;
        }
        System.out.println(Arrays.toString(array));
    }
    
}

File: python/ordenacao/InsertionSort.py
"""
Estruturas de Dados e Algoritmos
__author__ Rafael de Arruda Sobral
"""

def insertion_sort(array):
  
  """
  The InsertionSort algorithm iterates over the array and assumes that the visited 
  positions are already sorted in ascending order, meaning it only needs to find the 
  right position for the current element and insert it there.
  """
  
  for i in range(0, len(array)):
       key = array[i]
       aux = i - 1
       while (aux >= 0 and array[aux] > key):
          array[aux + 1] = array[aux]
          aux -= 1
       array[aux + 1] = key

def main():

    """ Simula a entrada e a saída de dados. """

    entrada = input().split()
    lista = [int(e) for e in entrada]
    insertion_sort(lista)
    print(lista)

main()
